Doc updates for StringInfo

danmoseley opened the pull request:
Used doc porter tool.

Fixes https://github.com/dotnet/runtime/issues/54876

opbld30 commented:
Docs Build status updates of commit _[7dec613](https://github.com/danmoseley/dotnet-api-docs/commits/7dec613e9514e50329f5a704c469fd325e84fa13)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[xml/System.Globalization/StringInfo.xml](https://github.com/danmoseley/dotnet-api-docs/blob/upd.globalization/xml/System.Globalization/StringInfo.xml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo?view=dotnet-uwp-10.0&branch=pr-en-us-6962) |

For more details, please refer to the [build report](https://opbuilduserstoragepublic.blob.core.windows.net/report/2021%5C8%5C3%5C500c749d-de78-3d4a-00e4-269d599bf9ec%5CPullRequest%5C202108030446209001-6962%5Cworkflow_report.html?sv=2018-03-28&sr=b&sig=q64lyp%2BmvE3cvAGf0bR8o8nXa1FHnXOttoc7ntlbqak%3D&st=2021-08-03T06%3A02%3A45Z&se=2021-09-03T06%3A07%3A45Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/500c749d-de78-3d4a-00e4-269d599bf9ec?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching the docs.microsoft.com <a href="https://review.docs.microsoft.com/en-us/help/?branch=master">contributor guides</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

tarekgh commented on xml/System.Globalization/StringInfo.xml:
```diff
@@ -491,9 +491,9 @@ This example uses the <xref:System.Globalization.StringInfo.GetTextElementEnumer
-        <param name="str">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="str">The input span to analyze.</param>
```
>The input span to analyze. [](http://example.com/codeflow?start=26&length=26)

The input char span from which to get the first text element length.

danmoseley replied:
I ended up leaving this, because the summary explains what it does. If I repeat the summary on this parameter you just get to read both. Note there are many of these here eg `The input string to analyze`

tarekgh commented on xml/System.Globalization/StringInfo.xml:
```diff
@@ -491,9 +491,9 @@ This example uses the <xref:System.Globalization.StringInfo.GetTextElementEnumer
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="str">The input span to analyze.</param>
+        <summary>Returns the length of the first text element (extended grapheme cluster) that occurs in the input span.</summary>
```
>(extended grapheme cluster) [](http://example.com/codeflow?start=62&length=27)

this part can be confusing as the first element in the text could be not a grapheme cluster and can bee a normal single character.

Maybe remove this part and add extra sentence telling  the first element could be grapheme cluster or surrogate pair which can have length greater than one character.

tarekgh commented:
CC @GrabYourPitchforks to validate the comments I added to this review.

tarekgh reviewed (approved):
add a couple of comments. LGTM otherwise.

GrabYourPitchforks commented:
> this part can be confusing as the first element in the text could be not a grapheme cluster and can be a normal single character.

A grapheme cluster might consist of just a single char, just like how a string or a `ROS<char>` can have a 1-char length if you really want. I wonder if we should reword the grapheme cluster docs to clarify this?

tarekgh commented:
> A grapheme cluster might consist of just a single char, just like how a string or a ROS<char> can have a 1-char length if you really want. I wonder if we should reword the grapheme cluster docs to clarify this?

Anyone reading `grapheme cluster` will not understand that :-) few people familiar with that. Clarification would be super helpful. at least we add something in the remarks explaining what `grapheme cluster` means.

GrabYourPitchforks commented:
@tarekgh The _Remarks_ section at https://docs.microsoft.com/dotnet/api/system.globalization.stringinfo starts off with this:

> .NET defines a text element as a unit of text that is displayed as a single character, that is, a grapheme. A text element can be a base character, a surrogate pair, or a combining character sequence.

You thinking that we should perhaps pull this same text in to these docs?

tarekgh commented:
> You thinking that we should perhaps pull this same text in to these docs?

Yes, that will be helpful. I would prefer the definition `A grapheme cluster is a sequence of one or more Unicode code points that should be treated as a single unit` to avoid tying it to display behaviors only. But I am fine with either definition.

danmoseley commented:
Thanks both. Note that this is just mechanically produced from the sources. I will turn around and make a PR in dotnet/runtime to update this text then regenerate this PR.

In general, it's surprising how "ready" the XML doc text in the sources is. It's certainly valuable to get it "right" there first time, in the PR process, so that ideally these PR's need no comment.

tarekgh reviewed (approved):
Thanks @danmoseley. LGTM.

danmoseley commented:
Updated. I added the comment to all 5 methods in the code, but only 3 ported because the GetNextTextElement overloads already have remarks. Those remarks are even better, so maybe should port back at some point.

I wonder if there's a flag on the doc porter tool to flag when it bails on porting remarks because they're present and different.

opbld34 commented:
Docs Build status updates of commit _[2c1fd99](https://github.com/danmoseley/dotnet-api-docs/commits/2c1fd990c9556a5620cd87527bd159c3184bcfcf)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[xml/System.Globalization/StringInfo.xml](https://github.com/danmoseley/dotnet-api-docs/blob/upd.globalization/xml/System.Globalization/StringInfo.xml) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo?view=dotnet-uwp-10.0&branch=pr-en-us-6962) |

For more details, please refer to the [build report](https://opbuilduserstoragepublic.blob.core.windows.net/report/2021%5C8%5C6%5C500c749d-de78-3d4a-00e4-269d599bf9ec%5CPullRequest%5C202108061617310285-6962%5Cworkflow_report.html?sv=2018-03-28&sr=b&sig=Yhb0VXpUmwxKx50inEGhuzLM0cFxlqOlJgn6QIeYPC8%3D&st=2021-08-06T16%3A42%3A01Z&se=2021-09-06T16%3A47%3A01Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/500c749d-de78-3d4a-00e4-269d599bf9ec?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching the docs.microsoft.com <a href="https://review.docs.microsoft.com/en-us/help/?branch=master">contributor guides</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>
